### PR TITLE
HDDS-8765. OzoneManagerLock to use striped locks to improve performance.

### PR DESCRIPTION
**Pull request by duongkame**
## What changes were proposed in this pull request?

This PR aims to fix the performance bottleneck of `LockManager` in Ozone Manager as described in HDDS-8765. The solution is to replace `LockManager` with striped locks. 

Some implementation details in this PR:
1. We use Guava's [Stripe](https://guava.dev/releases/19.0/api/docs/com/google/common/util/concurrent/Striped.html) for striped locks. 
2. Each resource type, e.g. BUCKET, VOLUME, KEY... has its own Striped. This prevents any collision cross-resource.
3. For composite lock keys like "$volume/$bucket/$key", I introduced a cheaper way, aka. CompositeKey, to avoid concatenating strings. And this eliminates the need for those the optimization of `hash` interfaces, e.g. `acquireWriteHashLock`.

TODO: the use of Guava's Striped is still under consideration because:
1. It's still a `@Beta` API, until [recently](https://github.com/google/guava/issues/3253). 
2. It [does not allow](https://github.com/google/guava/issues/2514) creating striped locks with `fair=true`. 

Unfortunately, there's no viable alternative. We may consider implementing our own Striped locks if (1) our policy doesn't allow `@Beta` APIs or (2) fair lock order is crucial.

### Performance evaluation
Did a pure read stress test (getKeyInfo) on an internal cluster. 
```
ozone freon ockrw -v duong1 -b obs1 --contiguous -n 30000000 -p zerobytes --percentage-read=100 -r 1000000 -s 0 --size 0 -t 150
```
On a cluster with 80 cores in OM and 100 RPC handler threads, this change improves the total OPPS from **105K** to **175K**.

Below are the detailed metrics. Note that for an OM request, read locks are acquired and released multiple times.

| Metric  | Before | After  | 
| ------------- | ------------- | ------------- |
| OPPS  | 105K  | 175K |
| getKeyInfo latency (μs)  | 875 | 380 |
| Avg lock acquiring (μs) | 65 | 0.06 |
| Avg Lock releasing (μs) | 65 | 1 |

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-8765

## How was this patch tested?

CI, existing unit tests.

### CODE REVIEW

**Comment by duongkame** [comment]
cc. @tanvipenumudy

**Review by adoroszlai — COMMENTED** [review]
Thanks @duongkame for the analysis in HDDS-8765 and the patch.  Mostly LGTM, some minor improvements suggested.

Ozone recently upgraded to Guava 32 where `Striped` is no longer `@Beta`.

You can get rid of some CI failures by merging `master` into your branch.

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/lock/OzoneManagerLock.java [hunk 1]
```diff
@@ -95,34 +97,35 @@ public class OzoneManagerLock implements IOzoneManagerLock {
...
+      ConfigurationSource conf) {
+    String stripeSizeKey = OZONE_MANAGER_STRIPED_LOCK_SIZE_PREFIX +
+        r.name().toLowerCase();
+    int size = conf.getInt(stripeSizeKey,
+        OZONE_MANAGER_STRIPED_LOCK_SIZE_DEFAULT);
+    return Striped.readWriteLock(size);
+  }
+
+  private ReentrantReadWriteLock getLock(Resource resource, String... keys) {
+    Object key = keys.length == 1 ? keys[0] : combineKeys(keys);
```
[the remark this answers is not in the document]
Comment (duongkame) [reply hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/lock/OzoneManagerLock.java]: Good one, thanks.

**Review by adoroszlai — COMMENTED** [review]
Thanks @duongkame for updating the patch, LGTM.

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/OzoneConfigKeys.java [hunk 2]
```diff
@@ -522,6 +522,11 @@ public final class OzoneConfigKeys {
   public static final String OZONE_MANAGER_FAIR_LOCK = "ozone.om.lock.fair";
   public static final boolean OZONE_MANAGER_FAIR_LOCK_DEFAULT = false;
 
+  public static final String OZONE_MANAGER_STRIPED_LOCK_SIZE_PREFIX =
+      "ozone.om.lock.striped.size.";
```
Comment (adoroszlai) [inline hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/OzoneConfigKeys.java:526]: When working on HDDS-8910 I found Ozone already uses `Striped` in a few places, and for `ContainerCache` the number of stripes is configurable:

https://github.com/apache/ozone/blob/98abce2e62396fcae554015d483d9d2c0fe0bb05/hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/OzoneConfigKeys.java#L164-L165

I think we should follow this naming convention for consistency, so I'd like to suggest a small change here:

```suggestion
      "ozone.om.lock.stripes.";
```

**Comment by adoroszlai** [comment]
@duongkame #5020 is also waiting for this to be merged, can we do so now, or would you like to run more tests with this?

**Comment by duongkame** [comment]
> @duongkame #5020 is also waiting for this to be merged, can we do so now, or would you like to run more tests with this?

I actually just need a +1 to merge this.

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/lock/OmReadOnlyLock.java [hunk 3]
```diff
@@ -17,48 +17,25 @@
 
 package org.apache.hadoop.ozone.om.lock;
 
+import org.apache.hadoop.ozone.om.lock.OzoneManagerLock.Resource;
+
 /**
  * Read only "lock" for snapshots
  * Uses no lock.  Always returns true when aquiring
```
Comment (kerneltime) [inline hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/lock/OmReadOnlyLock.java:24]:
```suggestion
 * Uses no lock.  Always returns true when acquiring
```

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/lock/OmReadOnlyLock.java [hunk 4]
```diff
@@ -17,48 +17,25 @@
...
  * read lock and false for write locks
  */
 public class OmReadOnlyLock implements IOzoneManagerLock {
```
Comment (kerneltime) [inline hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/lock/OmReadOnlyLock.java:27]: not related to this PR but this class should be renamed to something akin to `OmDummyReadLock`

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/lock/OzoneManagerLock.java [hunk 5]
```diff
@@ -95,34 +98,36 @@ public class OzoneManagerLock implements IOzoneManagerLock {
...
+    //  ReadWriteLock with fair-lock yet.
+    //  https://github.com/google/guava/issues/2514.
+    //  We may have to consider implement our own striped locks.
+    //boolean fair = conf.getBoolean(OZONE_MANAGER_FAIR_LOCK,
+    //    OZONE_MANAGER_FAIR_LOCK_DEFAULT);
+
+    Map<Resource, Striped<ReadWriteLock>> stripedLockMap =
+        new EnumMap<>(Resource.class);
+    for (Resource r : Resource.values()) {
+      stripedLockMap.put(r, createStripeLock(r, conf));
```
Comment (kerneltime) [inline hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/lock/OzoneManagerLock.java:113]: Any thoughts about setting the `fair` flag?

Comment (kerneltime) [reply hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/lock/OzoneManagerLock.java:113]: The `Striped` class for a single use should be fairly straighforward to mimic in our own code base.

**Comment by kerneltime** [comment]
The changes look good to me. I will give it one more look tomorrow. We can follow up on fairness in a separate PR.

**Comment by duongkame** [comment]
> The changes look good to me. I will give it one more look tomorrow. We can follow up on fairness in a separate PR.

Thanks Ritesh. Striped lock fairness will be addressed by HDDS-9237.

File: hadoop-ozone/common/src/test/java/org/apache/hadoop/ozone/om/lock/TestKeyPathLock.java [hunk 6]
```diff
@@ -218,26 +215,15 @@ private void testDiffKeyPathWriteLockMultiThreadingUtil(
       OzoneManagerLock lock, CountDownLatch countDown,
       String[] sampleResourceName) {
 
-    lock.acquireWriteHashedLock(resource,
-        generateResourceHashCode(resource, sampleResourceName));
+    lock.acquireWriteLock(resource, sampleResourceName);
     LOG.info("Write Lock Acquired by " + Thread.currentThread().getName());
 
     // Waiting for all the threads to be instantiated/to reach
     // acquireWriteHashedLock.
```
Comment (tanvipenumudy) [inline hadoop-ozone/common/src/test/java/org/apache/hadoop/ozone/om/lock/TestKeyPathLock.java:222]:
```suggestion
    // acquireWriteLock.
```

File: hadoop-ozone/common/src/test/java/org/apache/hadoop/ozone/om/lock/TestKeyPathLock.java [hunk 7]
```diff
@@ -218,26 +215,15 @@ private void testDiffKeyPathWriteLockMultiThreadingUtil(
...
     countDown.countDown();
-    while (countDown.getCount() > 0) {
```
Comment (tanvipenumudy) [inline hadoop-ozone/common/src/test/java/org/apache/hadoop/ozone/om/lock/TestKeyPathLock.java:228]: Nit: can we add `countDown.await()` here

File: hadoop-ozone/common/src/test/java/org/apache/hadoop/ozone/om/lock/TestKeyPathLock.java [hunk 8]
```diff
@@ -136,8 +136,7 @@ private void testSameKeyPathWriteLockMultiThreadingUtil(
     // Now all threads have been instantiated.
     Assert.assertEquals(0, countDownLatch.getCount());
 
-    lock.acquireWriteHashedLock(resource,
-        generateResourceHashCode(resource, sampleResourceName));
+    lock.acquireWriteLock(resource, sampleResourceName);
```
Comment (tanvipenumudy) [inline hadoop-ozone/common/src/test/java/org/apache/hadoop/ozone/om/lock/TestKeyPathLock.java:139]: Nit: Similar comment change under `TestKeyPathLock#testSameKeyPathWriteLockMultiThreadingUtil`:
https://github.com/apache/ozone/blob/1f57ff6ea7af532fdfe5fc8b1a11a5c53f744e43/hadoop-ozone/common/src/test/java/org/apache/hadoop/ozone/om/lock/TestKeyPathLock.java#L124-L125

Comment (tanvipenumudy) [inline hadoop-ozone/common/src/test/java/org/apache/hadoop/ozone/om/lock/TestKeyPathLock.java:139]: Nit: can we add `countDown.await()` similar to what is done under `TestKeyPathLock#testDiffKeyPathWriteLockMultiThreadingUtil`:
https://github.com/apache/ozone/blob/1f57ff6ea7af532fdfe5fc8b1a11a5c53f744e43/hadoop-ozone/common/src/test/java/org/apache/hadoop/ozone/om/lock/TestKeyPathLock.java#L127-L134

**Review by tanvipenumudy — COMMENTED** [review]
Thank you for the patch @duongkame, LGTM! 

I have just added some minor comments related to the tests.

File: hadoop-ozone/common/src/test/java/org/apache/hadoop/ozone/om/lock/TestKeyPathLock.java (hunk 7 above)
Comment (duongkame) [reply hadoop-ozone/common/src/test/java/org/apache/hadoop/ozone/om/lock/TestKeyPathLock.java:228]: For this aged PR, I think we should keep it as light as it is now. Any improvements in the existing test can be done separately.

Comment (tanvipenumudy) [reply hadoop-ozone/common/src/test/java/org/apache/hadoop/ozone/om/lock/TestKeyPathLock.java:228]: Sure, thanks @duongkame!

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/OzoneConfigKeys.java [hunk 9]
```diff
@@ -522,6 +522,11 @@ public final class OzoneConfigKeys {
   public static final String OZONE_MANAGER_FAIR_LOCK = "ozone.om.lock.fair";
   public static final boolean OZONE_MANAGER_FAIR_LOCK_DEFAULT = false;
 
+  public static final String OZONE_MANAGER_STRIPED_LOCK_SIZE_PREFIX =
+      "ozone.om.lock.stripes.";
```
Comment (kerneltime) [inline hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/OzoneConfigKeys.java:526]: Minor nit: ozone.om.lock.stripes.count

**Review by kerneltime — APPROVED** [review]
The change looks good.

File: hadoop-ozone/common/src/test/java/org/apache/hadoop/ozone/om/lock/TestKeyPathLock.java (hunk 8 above)
Comment (duongkame) [reply hadoop-ozone/common/src/test/java/org/apache/hadoop/ozone/om/lock/TestKeyPathLock.java:139]: similar to the above.

Comment (duongkame) [reply hadoop-ozone/common/src/test/java/org/apache/hadoop/ozone/om/lock/TestKeyPathLock.java:139]: similar to the above.

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/lock/OmReadOnlyLock.java (hunk 4 above)
Comment (duongkame) [reply hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/lock/OmReadOnlyLock.java:27]: Will do it in the next PR, this one is aged.

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/OzoneConfigKeys.java (hunk 9 above)
Comment (duongkame) [reply hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/OzoneConfigKeys.java:526]: Done.

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/OzoneConfigKeys.java [hunk 10]
```diff
@@ -519,6 +519,11 @@ public final class OzoneConfigKeys {
   public static final String OZONE_MANAGER_FAIR_LOCK = "ozone.om.lock.fair";
   public static final boolean OZONE_MANAGER_FAIR_LOCK_DEFAULT = false;
 
+  public static final String OZONE_MANAGER_STRIPED_LOCK_SIZE_PREFIX =
+      "ozone.om.lock.stripes.count.";
```
Comment (kerneltime) [inline hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/OzoneConfigKeys.java:523]:
```suggestion
      "ozone.om.lock.stripes.count";
```

Comment (adoroszlai) [reply hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/OzoneConfigKeys.java:523]: I would prefer `ozone.om.lock.stripes` (without `.count`), see https://github.com/apache/ozone/pull/4885#discussion_r1239365508.

And this is a prefix, so `.` is needed at the end.  Full config key is derived by appending resource name.

**Comment by duongkame** [comment]
After rebasing, a flaky test `TestOzoneFileSystem.testRenameFileToDir` always fails despite multiple retries. Seems to be a known issue. HDDS-9219.

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/OzoneConfigKeys.java (hunk 10 above)
Comment (duongkame) [reply hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/OzoneConfigKeys.java]: Thanks for the reminder @adoroszlai. I even forgot about that conversation. Updated.

**Pull request merged**